drivers: gpio: Fix GPIO initial status

benson0715 opened the pull request:
The current procedure to initialize the GPIO is:
If the GPIO is configured as GPIO_OUTPUT and also has the
GPIO_OUTPUT_INIT_HIGH flag, set it to output high; otherwise, set it to
output low.

This may fail if the GPIO is simply configured as GPIO_OUTPUT without
specifying either INIT_HIGH or INIT_LOW, which means it is intended to
preserve the previous status. But in this case, we are currently setting it
to output low.

Fix this by explicitly initializing the GPIO to high or low according to the
configuration:
If the GPIO is configured as GPIO_OUTPUT and also with the
GPIO_OUTPUT_INIT_HIGH flag, set it to output high.  
If the GPIO is configured as GPIO_OUTPUT and also with the
GPIO_OUTPUT_INIT_LOW flag, set it to output low.
If the GPIO is configured as GPIO_OUTPUT only, do not set the output value.

henrikbrixandersen reviewed (changes requested):
> ```
> During jumping, GPIO initialization may not performed
> in order to preserve the previous status.
> But, original implementation would default to
> output low regardless of the flag.
> So, fix it.
> ```

I must admit I do not fully understand your PR description/Git commit log. Could you please try rephrasing it? You might be able to find inspiration in the Git commit logs in the repo history.

benson0715 commented:
> > ```
> > During jumping, GPIO initialization may not performed
> > in order to preserve the previous status.
> > But, original implementation would default to
> > output low regardless of the flag.
> > So, fix it.
> > ```
> 
> I must admit I do not fully understand your PR description/Git commit log. Could you please try rephrasing it? You might be able to find inspiration in the Git commit logs in the repo history.

I rephrased my commit log, thanks!

benson0715 committed:
drivers: gpio: Fix GPIO initial status

The current procedure to initialize the GPIO is:
If the GPIO is configured as GPIO_OUTPUT and also has the
GPIO_OUTPUT_INIT_HIGH flag, set it to output high; otherwise, set it to
output low.

This may fail if the GPIO is simply configured as GPIO_OUTPUT without
specifying either INIT_HIGH or INIT_LOW, which means it is intended to
preserve the previous status. But in this case, we are currently setting it
to output low.

Fix this by explicitly initializing the GPIO to high or low according to
the configuration:
If the GPIO is configured as GPIO_OUTPUT and also with the
GPIO_OUTPUT_INIT_HIGH flag, set it to output high.
If the GPIO is configured as GPIO_OUTPUT and also with the
GPIO_OUTPUT_INIT_LOW flag, set it to output low.
If the GPIO is configured as GPIO_OUTPUT only,
do not set the output value.

Signed-off-by: Benson Huang <benson7633769@gmail.com>
